challenges/security/application-security.md

```diff
@@ -0,0 +1,245 @@
+
+* Do not implement a system that scales or is highly performing. Describe which
+  performance improvements you would add in the future.
+* It is OK if the system is not highly available. Write down how you would make
+  the system highly available and why your system is not.
+
+## Pitfalls and Gotchas
+
+To help you out, we've composed a list of things that previously resulted in a
+no-pass from the interview team:
+
+* Scope creep. Candidates have tried to implement too much and ran out of time
+  and energy. To avoid this pitfall, use the simplest solution that will work.
+  Avoid writing too much code. We've seen candidates' code introducing caching
+  and making many mistakes in the caching layer validation logic. Not having
+  caching would have solved this problem.
+* Data races. We will scan the code with a race detector and do our best to find
+  data races in the code. Avoid global state as much as possible; if using
+  global state, write down a good description why it is necessary and protect it
+  against data races.
+* Deadlocks. When using mutexes, channels or any other synchronization
+  primitives, make sure the system won't deadlock. We've seen candidates' code
+  holding a mutex and making a network call without timeouts in place. Be extra
+  careful with networking and sync primitives.
+* Unstructured code. We've seen candidates leaving commented chunks of code,
+  having one large file with all the code, not having code structure at all.
+* Not communicating. Some candidates have submitted all their code to the master
+  branch without raising pull requests, which does not give us the ability to
+  provide feedback on the various implementation phases. We are a distributed
+  team, so structured, asynchronous communication is critical to us.
+
+## Questions
+
+It is OK to ask the interview team questions. Some folks stay away from asking
+questions to avoid appearing less experienced, so we provide examples of
+questions to ask and questions we expect candidates to figure out on their own.
+
+Here is a great question to ask:
+
+> Is it OK to pre-generate secret data and put the secrets in the repository for
+> a proof of concept? I will add a note that we will auto-generate secrets in
+> the future.
+
+It demonstrates that you thought about this problem domain, recognize the trade
+off and are saving you and the team time by not implementing it.
+
+This is the question we expect candidates to figure out on their own:
+
+> Which logging library should I use?
+
+Unless specified in the requirements, pick the solution that works best for you.
+
+# Tools
+
+The prototype codebase will be provided in go and the final version should build
+and run on 64-bit Linux machines.
+
+# Timing
+
+You can split coding over a couple of weekdays or weekends and find time to ask
+questions and receive feedback.
+
+Once you join the Slack channel, you have between 1 week complete the challenge.
```

Comment:
```suggestion
Once you join the Slack channel, you have 1 week complete the challenge.
```

We're trying to keep this from creeping to a 2 week challenge.  Ideally, we'd have a it be a one day challenge, but we're not there yet.

Reply:
This was rewritten to be a one day challenge in https://github.com/gravitational/careers/pull/66/commits/bb7b1e0b02198ccfdf0d1d6d91e9dcac42818586